base/dict.jl

```diff
@@ -391,6 +391,15 @@ function setindex!(h::Dict{K,V}, v0, key::K) where V where K
     return h
 end
 
+function setindex(d0::Dict, v, k)
+    K = promote_type(keytype(d0), typeof(k))
+    V = promote_type(valtype(d0), typeof(v))
+    d = Dict{K, V}()
```

Comment:
As an optimization it might be useful to do `if typeof(d0) == Dict{K,V}; id = copy(d0)`. Or does `copy(d, d0)` call an optimized method which doesn't reconstruct the dict already?

Reply:
tl;dr: It does not seem that there is a special method we can use for copying a `Dict`.

---

It looks like Base only has a basic definition for `copy!` of dicts:

https://github.com/JuliaLang/julia/blob/7bc374290cbb62ff1890541226da2df8fb3ef864/base/abstractdict.jl#L153-L154

`copy(::Dict)` has a specialization:

https://github.com/JuliaLang/julia/blob/7bc374290cbb62ff1890541226da2df8fb3ef864/base/dict.jl#L119

But this `Dict(::Dict)` dispatches to a generic fallback:

```julia
julia> methods(Dict, Tuple{Dict})
# 1 method for type constructor:
[1] (::Type{Dict})(kv) in Base at dict.jl:127
```

which is defined as

https://github.com/JuliaLang/julia/blob/7bc374290cbb62ff1890541226da2df8fb3ef864/base/dict.jl#L126-L128

`dict_with_eltype` does not seem to have a specialization for `Dict`:

```julia
julia> methods(Base.dict_with_eltype)
# 6 methods for generic function "dict_with_eltype":
[1] dict_with_eltype(DT_apply, kv::Base.Generator, ::Union{Type{Pair{K,V}}, Type{Tuple{K,V}}}) where {K, V} in Base at abstractdict.jl:540
[2] dict_with_eltype(DT_apply, ::Type{Pair{K,V}}) where {K, V} in Base at abstractdict.jl:541
[3] dict_with_eltype(DT_apply, ::Type) in Base at abstractdict.jl:542
[4] dict_with_eltype(DT_apply, kv, ::Union{Type{Pair{K,V}}, Type{Tuple{K,V}}}) where {K, V} in Base at abstractdict.jl:539
[5] dict_with_eltype(DT_apply::F, kv::Base.Generator, t) where F in Base at abstractdict.jl:545
[6] dict_with_eltype(DT_apply::F, kv, t) where F in Base at abstractdict.jl:543
```

Reply:
I think you missed this inner constructor:
https://github.com/JuliaLang/julia/blob/7bc374290cbb62ff1890541226da2df8fb3ef864/base/dict.jl#L92

Reply:
Thanks, I missed that. Indeed, `copy(::Dict)` in the end hits `Dict{K,V}(d::Dict{K,V})`.  But maybe we can do this in another PR?  It is quite easy to add a specialization for `copy!` (so that it's usable outside `setindex`) that works well for different value types:

```julia
function copy!(dst::Dict{K}, src::Dict{K}) where K
    empty!(dst)
    append!(dst.vals, src.vals)  # try first, as it may fail
    append!(dst.keys, src.keys)
    append!(dst.slots, src.slots)
    dst.ndel = src.ndel
    dst.count = src.count
    dst.age = src.age
    dst.idxfloor = src.idxfloor
    dst.maxprobe = src.maxprobe
    return dst
end
```

Reply:
I opened a PR for optimizing `copy!(::Dict, ::Dict)` https://github.com/JuliaLang/julia/pull/34101